Ubuntu: https://bugs.launchpad.net/ubuntu/+source/grub2/+bug/216178
Upstream: http://lists.gnu.org/archive/html/grub-devel/2009-09/msg00040.html
 and thread
Description: Add GRUB_DEFAULT=saved, as well as grub-set-default and
 grub-reboot utilities. Provides functionality essentially equivalent to
 GRUB Legacy's savedefault.

diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/conf/common.rmk grub2-1.97~beta3.new/conf/common.rmk
--- grub2-1.97~beta3/conf/common.rmk	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/conf/common.rmk	2009-09-17 13:40:31.000000000 +0100
@@ -175,6 +175,20 @@
 sbin_SCRIPTS += grub-dumpbios
 CLEANFILES += grub-dumpbios
 
+# For grub-set-default.
+grub-set-default: util/grub-set-default.in config.status
+	./config.status --file=$@:$<
+	chmod +x $@
+sbin_SCRIPTS += grub-set-default
+CLEANFILES += grub-set-default
+
+# For grub-reboot.
+grub-reboot: util/grub-reboot.in config.status
+	./config.status --file=$@:$<
+	chmod +x $@
+sbin_SCRIPTS += grub-reboot
+CLEANFILES += grub-reboot
+
 # Filing systems.
 pkglib_MODULES += fshelp.mod fat.mod ufs1.mod ufs2.mod ext2.mod ntfs.mod \
 	ntfscomp.mod minix.mod hfs.mod jfs.mod iso9660.mod xfs.mod	\
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/normal/menu.c grub2-1.97~beta3.new/normal/menu.c
--- grub2-1.97~beta3/normal/menu.c	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/normal/menu.c	2009-09-17 13:40:31.000000000 +0100
@@ -137,6 +137,8 @@
       return;
     }
 
+  grub_env_set ("chosen", entry->title);
+
   grub_parser_execute ((char *) entry->sourcecode);
 
   if (grub_errno == GRUB_ERR_NONE && grub_loader_is_loaded ())
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/normal/menu_text.c grub2-1.97~beta3.new/normal/menu_text.c
--- grub2-1.97~beta3/normal/menu_text.c	2009-09-17 13:40:24.000000000 +0100
+++ grub2-1.97~beta3.new/normal/menu_text.c	2009-09-17 13:40:31.000000000 +0100
@@ -237,7 +237,7 @@
 
 /* Get the entry number from the variable NAME.  */
 static int
-get_entry_number (const char *name)
+get_entry_number (grub_menu_t menu, const char *name)
 {
   char *val;
   int entry;
@@ -250,6 +250,28 @@
 
   entry = (int) grub_strtoul (val, 0, 0);
 
+  if (grub_errno == GRUB_ERR_BAD_NUMBER)
+    {
+      /* See if the variable matches the title of a menu entry.  */
+      grub_menu_entry_t e = menu->entry_list;
+      int i;
+
+      grub_errno = GRUB_ERR_NONE;
+
+      for (i = 0; e; i++)
+	{
+	  if (grub_strcmp (e->title, val) == 0)
+	    {
+	      entry = i;
+	      break;
+	    }
+	  e = e->next;
+	}
+
+      if (! e)
+	entry = -1;
+    }
+
   if (grub_errno != GRUB_ERR_NONE)
     {
       grub_errno = GRUB_ERR_NONE;
@@ -291,7 +313,7 @@
 
   first = 0;
 
-  default_entry = get_entry_number ("default");
+  default_entry = get_entry_number (menu, "default");
 
   /* If DEFAULT_ENTRY is not within the menu entries, fall back to
      the first entry.  */
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub-mkconfig_lib.in grub2-1.97~beta3.new/util/grub-mkconfig_lib.in
--- grub2-1.97~beta3/util/grub-mkconfig_lib.in	2009-09-17 13:40:23.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub-mkconfig_lib.in	2009-09-17 13:40:31.000000000 +0100
@@ -130,6 +130,14 @@
   echo ${drive}${relative_path}
 }
 
+save_default_entry ()
+{
+  if [ "x${GRUB_DEFAULT}" = "xsaved" ] ; then
+    echo 'saved_entry=${chosen}'
+    echo 'save_env saved_entry'
+  fi
+}
+
 prepare_grub_to_access_device ()
 {
   device=$1
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub-reboot.in grub2-1.97~beta3.new/util/grub-reboot.in
--- grub2-1.97~beta3/util/grub-reboot.in	1970-01-01 01:00:00.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub-reboot.in	2009-09-17 13:40:31.000000000 +0100
@@ -0,0 +1,104 @@
+#! /bin/sh
+#
+# Set a default boot entry for GRUB, for the next boot only.
+# Copyright (C) 2004,2009  Free Software Foundation, Inc.
+#
+# GRUB is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# GRUB is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
+
+# Initialize some variables.
+transform="@program_transform_name@"
+
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+bindir=@bindir@
+
+grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
+rootdir=
+
+# Usage: usage
+# Print the usage.
+usage () {
+    cat <<EOF
+Usage: $0 [OPTION] entry
+Set the default boot entry for GRUB, for the next boot only.
+
+  -h, --help              print this message and exit
+  -v, --version           print the version information and exit
+  --root-directory DIR    expect GRUB images under the directory DIR
+                          instead of the root directory
+
+ENTRY is a number or a menu item title.
+
+Report bugs to <bug-grub@gnu.org>.
+EOF
+}
+
+# Check the arguments.
+for option in "$@"; do
+    case "$option" in
+    -h | --help)
+	usage
+	exit 0 ;;
+    -v | --version)
+	echo "grub-install (GNU GRUB ${PACKAGE_VERSION})"
+	exit 0 ;;
+    --root-directory=*)
+	rootdir=`echo "$option" | sed 's/--root-directory=//'` ;;
+    -*)
+	echo "Unrecognized option \`$option'" 1>&2
+	usage
+	exit 1
+	;;
+    *)
+	if test "x$entry" != x; then
+	    echo "More than one entry?" 1>&2
+	    usage
+	    exit 1
+	fi
+	entry="${option}" ;;
+    esac
+done
+
+if test "x$entry" = x; then
+    echo "entry not specified." 1>&2
+    usage
+    exit 1
+fi
+
+# Initialize these directories here, since ROOTDIR was initialized.
+case "$host_os" in
+netbsd* | openbsd*)
+    # Because /boot is used for the boot block in NetBSD and OpenBSD, use /grub
+    # instead of /boot/grub.
+    grub_prefix=`echo /grub | sed ${transform}`
+    bootdir=${rootdir}
+    ;;
+*)
+    # Use /boot/grub by default.
+    bootdir=${rootdir}/boot
+    ;;
+esac
+
+grubdir=${bootdir}/`echo grub | sed ${transform}`
+
+prev_saved_entry=`$grub_editenv ${grubdir}/grubenv list | sed -n 's/^saved_entry=//p'`
+if [ "$prev_saved_entry" ]; then
+    $grub_editenv ${grubdir}/grubenv set prev_saved_entry="$prev_saved_entry"
+else
+    $grub_editenv ${grubdir}/grubenv unset prev_saved_entry
+fi
+$grub_editenv ${grubdir}/grubenv set saved_entry="$entry"
+
+# Bye.
+exit 0
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub-set-default.in grub2-1.97~beta3.new/util/grub-set-default.in
--- grub2-1.97~beta3/util/grub-set-default.in	1970-01-01 01:00:00.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub-set-default.in	2009-09-17 13:40:31.000000000 +0100
@@ -0,0 +1,99 @@
+#! /bin/sh
+#
+# Set a default boot entry for GRUB.
+# Copyright (C) 2004,2009  Free Software Foundation, Inc.
+#
+# GRUB is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# GRUB is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
+
+# Initialize some variables.
+transform="@program_transform_name@"
+
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+bindir=@bindir@
+
+grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
+rootdir=
+
+# Usage: usage
+# Print the usage.
+usage () {
+    cat <<EOF
+Usage: $0 [OPTION] entry
+Set the default boot entry for GRUB.
+
+  -h, --help              print this message and exit
+  -v, --version           print the version information and exit
+  --root-directory DIR    expect GRUB images under the directory DIR
+                          instead of the root directory
+
+ENTRY is a number or a menu item title.
+
+Report bugs to <bug-grub@gnu.org>.
+EOF
+}
+
+# Check the arguments.
+for option in "$@"; do
+    case "$option" in
+    -h | --help)
+	usage
+	exit 0 ;;
+    -v | --version)
+	echo "grub-install (GNU GRUB ${PACKAGE_VERSION})"
+	exit 0 ;;
+    --root-directory=*)
+	rootdir=`echo "$option" | sed 's/--root-directory=//'` ;;
+    -*)
+	echo "Unrecognized option \`$option'" 1>&2
+	usage
+	exit 1
+	;;
+    *)
+	if test "x$entry" != x; then
+	    echo "More than one entry?" 1>&2
+	    usage
+	    exit 1
+	fi
+	entry="${option}" ;;
+    esac
+done
+
+if test "x$entry" = x; then
+    echo "entry not specified." 1>&2
+    usage
+    exit 1
+fi
+
+# Initialize these directories here, since ROOTDIR was initialized.
+case "$host_os" in
+netbsd* | openbsd*)
+    # Because /boot is used for the boot block in NetBSD and OpenBSD, use /grub
+    # instead of /boot/grub.
+    grub_prefix=`echo /grub | sed ${transform}`
+    bootdir=${rootdir}
+    ;;
+*)
+    # Use /boot/grub by default.
+    bootdir=${rootdir}/boot
+    ;;
+esac
+
+grubdir=${bootdir}/`echo grub | sed ${transform}`
+
+$grub_editenv ${grubdir}/grubenv unset prev_saved_entry
+$grub_editenv ${grubdir}/grubenv set saved_entry="$entry"
+
+# Bye.
+exit 0
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub.d/00_header.in grub2-1.97~beta3.new/util/grub.d/00_header.in
--- grub2-1.97~beta3/util/grub.d/00_header.in	2009-09-17 13:40:24.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub.d/00_header.in	2009-09-17 13:43:50.000000000 +0100
@@ -32,11 +32,19 @@
 done
 
 if [ "x${GRUB_DEFAULT}" = "x" ] ; then GRUB_DEFAULT=0 ; fi
+if [ "x${GRUB_DEFAULT}" = "xsaved" ] ; then GRUB_DEFAULT='${saved_entry}' ; fi
 if [ "x${GRUB_TIMEOUT}" = "x" ] ; then GRUB_TIMEOUT=5 ; fi
 if [ "x${GRUB_GFXMODE}" = "x" ] ; then GRUB_GFXMODE=640x480 ; fi
 
 cat << EOF
-set default=${GRUB_DEFAULT}
+load_env
+set default="${GRUB_DEFAULT}"
+if [ \${prev_saved_entry} ]; then
+  saved_entry=\${prev_saved_entry}
+  save_env saved_entry
+  prev_saved_entry=
+  save_env prev_saved_entry
+fi
 EOF
 
 case ${GRUB_TERMINAL_INPUT}:${GRUB_TERMINAL_OUTPUT} in
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub.d/10_freebsd.in grub2-1.97~beta3.new/util/grub.d/10_freebsd.in
--- grub2-1.97~beta3/util/grub.d/10_freebsd.in	2009-09-17 13:40:22.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub.d/10_freebsd.in	2009-09-17 13:40:31.000000000 +0100
@@ -31,6 +31,7 @@
   cat << EOF
 menuentry "$1" {
 EOF
+  save_default_entry | sed -e "s/^/\t/"
   prepare_grub_to_access_device ${GRUB_DEVICE_BOOT} | sed -e "s/^/\t/"
   cat << EOF
 	freebsd			${rel_dirname}/${basename}
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub.d/10_hurd.in grub2-1.97~beta3.new/util/grub.d/10_hurd.in
--- grub2-1.97~beta3/util/grub.d/10_hurd.in	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub.d/10_hurd.in	2009-09-17 13:40:31.000000000 +0100
@@ -71,6 +71,7 @@
 cat << EOF
 menuentry "${OS}" {
 EOF
+save_default_entry | sed -e "s/^/\t/"
 prepare_grub_to_access_device ${GRUB_DEVICE} | sed -e "s/^/\t/"
 cat << EOF
 	multiboot ${kernel} root=device:${GRUB_DEVICE}
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub.d/10_linux.in grub2-1.97~beta3.new/util/grub.d/10_linux.in
--- grub2-1.97~beta3/util/grub.d/10_linux.in	2009-09-17 13:40:24.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub.d/10_linux.in	2009-09-17 13:40:31.000000000 +0100
@@ -58,6 +58,7 @@
 	set quiet=1
 EOF
   fi
+  save_default_entry | sed -e "s/^/\t/"
   prepare_grub_to_access_device ${GRUB_DEVICE_BOOT} | sed -e "s/^/\t/"
   cat << EOF
 	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro $2
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub.d/10_windows.in grub2-1.97~beta3.new/util/grub.d/10_windows.in
--- grub2-1.97~beta3/util/grub.d/10_windows.in	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub.d/10_windows.in	2009-09-17 13:40:31.000000000 +0100
@@ -73,6 +73,7 @@
 menuentry "$OS" {
 EOF
 
+  save_default_entry | sed -e 's,^,\t,'
   prepare_grub_to_access_device "$dev" | sed 's,^,\t,'
 
   cat << EOF
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/grub.d/30_os-prober.in grub2-1.97~beta3.new/util/grub.d/30_os-prober.in
--- grub2-1.97~beta3/util/grub.d/30_os-prober.in	2009-09-17 13:40:24.000000000 +0100
+++ grub2-1.97~beta3.new/util/grub.d/30_os-prober.in	2009-09-17 13:40:31.000000000 +0100
@@ -99,6 +99,7 @@
       cat << EOF
 menuentry "${LONGNAME} (on ${DEVICE})" {
 EOF
+      save_default_entry | sed -e "s/^/\t/"
       prepare_grub_to_access_device ${DEVICE} | sed -e "s/^/\t/"
 
       if [ "${LONGNAME}" != "Windows Vista (loader)" ] ; then
@@ -129,6 +130,7 @@
         cat << EOF
 menuentry "${LLABEL} (on ${DEVICE})" {
 EOF
+	save_default_entry | sed -e "s/^/\t/"
 	prepare_grub_to_access_device ${DEVICE} | sed -e "s/^/\t/"
 	cat <<  EOF
 	linux ${LKERNEL} ${LPARAMS}
@@ -148,6 +150,7 @@
         cat << EOF
 menuentry "${LONGNAME} (on ${DEVICE})" {
 EOF
+	save_default_entry | sed -e "s/^/\t/"
 	prepare_grub_to_access_device ${DEVICE} | sed -e "s/^/\t/"
 	cat << EOF
         insmod vbe
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/i386/efi/grub-install.in grub2-1.97~beta3.new/util/i386/efi/grub-install.in
--- grub2-1.97~beta3/util/i386/efi/grub-install.in	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/util/i386/efi/grub-install.in	2009-09-17 13:40:31.000000000 +0100
@@ -34,6 +34,7 @@
 grub_mkimage=${bindir}/`echo grub-mkimage | sed ${transform}`
 grub_mkdevicemap=${sbindir}/`echo grub-mkdevicemap | sed ${transform}`
 grub_probe=${sbindir}/`echo grub-probe | sed ${transform}`
+grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
 rootdir=
 grub_prefix=`echo /boot/grub | sed ${transform}`
 modules=
@@ -178,6 +179,10 @@
     cp -f $file ${grubdir} || exit 1
 done
 
+if ! test -f ${grubdir}/grubenv; then
+    $grub_editenv ${grubdir}/grubenv create
+fi
+
 # Create the core image. First, auto-detect the filesystem module.
 fs_module=`$grub_probe --target=fs --device-map=${device_map} ${grubdir}`
 if test "x$fs_module" = xfat; then :; else
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/i386/pc/grub-install.in grub2-1.97~beta3.new/util/i386/pc/grub-install.in
--- grub2-1.97~beta3/util/i386/pc/grub-install.in	2009-09-17 13:40:24.000000000 +0100
+++ grub2-1.97~beta3.new/util/i386/pc/grub-install.in	2009-09-17 13:40:31.000000000 +0100
@@ -39,6 +39,7 @@
 fi
 grub_mkdevicemap=${sbindir}/`echo grub-mkdevicemap | sed ${transform}`
 grub_probe=${sbindir}/`echo grub-probe | sed ${transform}`
+grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
 rootdir=
 grub_prefix=`echo /boot/grub | sed ${transform}`
 modules=
@@ -250,6 +251,10 @@
     done
 fi
 
+if ! test -f ${grubdir}/grubenv; then
+    $grub_editenv ${grubdir}/grubenv create
+fi
+
 # Write device to a variable so we don't have to traverse /dev every time.
 grub_device=`$grub_probe --target=device ${grubdir}`
 
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/ieee1275/grub-install.in grub2-1.97~beta3.new/util/ieee1275/grub-install.in
--- grub2-1.97~beta3/util/ieee1275/grub-install.in	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/util/ieee1275/grub-install.in	2009-09-17 13:40:31.000000000 +0100
@@ -37,6 +37,7 @@
 grub_mkimage=${bindir}/`echo grub-mkelfimage | sed ${transform}`
 grub_mkdevicemap=${sbindir}/`echo grub-mkdevicemap | sed ${transform}`
 grub_probe=${sbindir}/`echo grub-probe | sed ${transform}`
+grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
 rootdir=
 grub_prefix=`echo /boot/grub | sed ${transform}`
 modules=
@@ -163,6 +164,10 @@
     cp -f $file ${grubdir} || exit 1
 done
 
+if ! test -f ${grubdir}/grubenv; then
+    $grub_editenv ${grubdir}/grubenv create
+fi
+
 # Create the core image. First, auto-detect the filesystem module.
 fs_module=`$grub_probe --target=fs --device-map=${device_map} ${grubdir}`
 if test "x$fs_module" = x -a "x$modules" = x; then
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta3/util/sparc64/ieee1275/grub-install.in grub2-1.97~beta3.new/util/sparc64/ieee1275/grub-install.in
--- grub2-1.97~beta3/util/sparc64/ieee1275/grub-install.in	2009-09-12 14:29:38.000000000 +0100
+++ grub2-1.97~beta3.new/util/sparc64/ieee1275/grub-install.in	2009-09-17 13:40:31.000000000 +0100
@@ -35,6 +35,7 @@
 grub_mkimage=${bindir}/`echo grub-mkimage | sed ${transform}`
 grub_mkdevicemap=${sbindir}/`echo grub-mkdevicemap | sed ${transform}`
 grub_probe=${sbindir}/`echo grub-probe | sed ${transform}`
+grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
 rootdir=
 grub_prefix=`echo /boot/grub | sed ${transform}`
 modules=
@@ -206,6 +207,10 @@
     cp -f $file ${grubdir} || exit 1
 done
 
+if ! test -f ${grubdir}/grubenv; then
+    $grub_editenv ${grubdir}/grubenv create
+fi
+
 # Write device to a variable so we don't have to traverse /dev every time.
 grub_device=`$grub_probe --target=device ${grubdir}`
 
